scripts/install-mender.sh

```diff
@@ -0,0 +1,147 @@
+#!/bin/sh
+
+set -e
+
+CHANNEL="stable"
+
+DEFAULT_COMPONENTS="mender-client"
+COMPONENTS="$DEFAULT_COMPONENTS mender-shell"
+SELECTED_COMPONENTS=$DEFAULT_COMPONENTS
+
+export DEBIAN_FRONTEND=noninteractive
+
+banner (){
+    echo "
+                          _
+ _ __ ___   ___ _ __   __| | ___ _ __ 
+| '_ \` _ \ / _ \ '_ \ / _\` |/ _ \ '__|
+| | | | | |  __/ | | | (_| |  __/ |   
+|_| |_| |_|\___|_| |_|\__,_|\___|_|   
+
+Running the Mender installation script.
+--
+"
+                                      
+}
+
+usage() {
+    echo "usage: install-mender.sh [-h] [-c channel] [component...]"
+    echo "  -h             print this help"
+    echo "  -c             channel: stable(default)|experimental"
+    echo "  <component>    list of components to install"
+    echo ""
+    echo "Supported components:"
+    echo "  * mender-client (default if none provided)"
+    echo "  * mender-shell"
+}
+
+is_known_component() {
+    for known in $COMPONENTS
+    do
+	if [ "$1" = "$known" ]	
+ 	then
+	  return 0 
+	fi
+    done
+    return 1
+}
+
+parse_args() {
+    while [ $# -gt 0 ]
+    do
+	case $1 in
+	    -h)
+		usage
+		exit 0
+ 	        ;;
+	    -c)
+	        if [ -n "$2" ]; then
+                    CHANNEL=$2
+	            shift 2
+		else
+                    echo "ERROR: channel requires a non-empty option argument."
+                    echo "Aborting."
+                    exit 1
+                fi
+		;;
+	    *)
+        	SELECTED_COMPONENTS=""
+	    	if is_known_component "$1"
+		    then
+			SELECTED_COMPONENTS="$SELECTED_COMPONENTS $1 "
+			shift
+		    else
+			echo "Unsupported argument: \`$1\`"
+			echo "Run \`mender-install.sh -h\` for help."
+			echo "Aborting."
+			exit 1
+		fi
+		;;
+	esac
+    done
+}
+
+print_components() {
+    echo "  Selected components:"
+    for c in $SELECTED_COMPONENTS
+    do
+        echo "\t$c" 
+    done
+}
+
+init() {
+    REPO_URL=https://downloads.mender.io/repos/debian
+
+    parse_args "$@"
+
+    ARCH=$(dpkg --print-architecture)
+    echo "  Detected architecture:"
+    echo "\t$ARCH"
+
+    echo "  Installing from channel:"
+    echo "\t$CHANNEL"
+}
+
+get_deps() {
+    apt-get update -qq
+    apt-get install -y -qq --no-install-recommends \
+        apt-transport-https \
+        ca-certificates \
+        curl \
+        gnupg-agent \
+        software-properties-common
+}
+
+add_repo() {
+    curl -fsSL $REPO_URL/gpg | apt-key add -
+
+    add-apt-repository \
+        "deb [arch=$ARCH] $REPO_URL \
+        $CHANNEL \
+        main"
+}
+
+inst_component() {
+    echo "Installing $1"
+    apt-get install -y -qq "$1"
+}
+
+do_install() {
+    apt-get update 
+
+    for c in $SELECTED_COMPONENTS
+    do
+        inst_component "$c"
```

Comment:
I think this will not work if the user happens to list the components in opposite order, because mender-shell depends on mender-client, which will not be installed until the second step. It should be fine to just install both in one go, apt supports multiple install targets.

Reply:
That is correct.

Also we have to experiment a bit on what happens if `mender-client` is already installed, which may happen 1) with a user that has installed an (assume old) package from Debian upstream or 2) a pre-build image which contains the binary, inventory scripts, etc but was not installed as a deb package (which is the case for onboarding).

For these, adding some force flags should do the job.

Reply:
Look at [GUI code](https://github.com/mendersoftware/gui/blob/f84425618f867476c95f401a4125d627220358fa/src/js/helpers.js#L430) for inspiration.

Reply:
I fixed the basic remark here: the list is now really a var, and everything is installed in one go.

I'll play some more with the force flags, these will come later.